### PR TITLE
Generate scheduling information statistics on right-click+drag

### PR DESCRIPTION
**Pull request by pierricgimmig**
When selecting a time window in the capture view, in devmode only,
we can now see a detailed breakdown of scheduling information in the
"Debug" tab, under "Selection Summary". This includes core occupancy
information as well as system-wide process and thread on-core time
statistics.

image

### CODE REVIEW

File: src/OrbitGl/CaptureStatsTest.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  scopes.push_back(CreateScope(/*pid=*/2, /*tid=*/3, /*cpu=*/6, /*start_ns=*/0, /*end_ns=*/100));
+  scopes.push_back(CreateScope(/*pid=*/2, /*tid=*/3, /*cpu=*/7, /*start_ns=*/0, /*end_ns=*/100));
+
+  {
+    SchedulingStats scheduling_stats(scopes, thread_name_provider, /*start_ns=*/0, /*end_ns*/ 1000);
+
+    EXPECT_EQ(scheduling_stats.GetTimeRangeMs(), 0.001);
+    EXPECT_EQ(scheduling_stats.GetTimeOnCoreNsByCore().size(), 8);
+    EXPECT_EQ(scheduling_stats.GetProcessStatsByPid().size(), 3);
+    EXPECT_EQ(scheduling_stats.GetProcessStatsSortedByTimeOnCore().size(), 3);
```
Comment (karupayun) [inline src/OrbitGl/CaptureStatsTest.cpp:97]: I would like to also check a little more about detailed information based on the the scopes you put before

File: src/OrbitGl/CaptureStats.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ErrorMessageOr<void> CaptureStats::Generate(CaptureWindow* capture_window, uint64_t start_ns,
+                                            uint64_t end_ns) {
+  ORBIT_SCOPE_FUNCTION;
+  if (capture_window == nullptr) return ErrorMessage("CaptureWindow is null");
+  if (start_ns == end_ns) return ErrorMessage("Time range is 0");
+  if (start_ns > end_ns) std::swap(start_ns, end_ns);
+
+  TimeGraph* time_graph = capture_window->GetTimeGraph();
+  TrackManager* track_manager = time_graph->GetTrackManager();
+  SchedulerTrack* scheduler_track = track_manager->GetOrCreateSchedulerTrack();
```
Comment (karupayun) [inline src/OrbitGl/CaptureStats.cpp:22]: nit: As we are using track_manager only once, why not 
`scheduler_track = time_graph->GetTrackManager()->GetOrCreateSchedulerTrack()`?

File: src/OrbitGl/SchedulingStats.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  for (auto& [core, time_on_core_ns_] : time_on_core_ns_by_core_) {
+    double time_on_core_ms = static_cast<double>(time_on_core_ns_) * kNsToMs;
+    summary +=
+        absl::StrFormat("cpu[%u] : %.2f%%\n", core, 100.0 * time_on_core_ms / time_range_ms_);
+  }
+
+  // Process and thread stats.
+  if (time_range_ms_ > 0) summary += absl::StrFormat("\nSelection time: %.6f ms\n", time_range_ms_);
+  for (ProcessStats* p_stats : process_stats_sorted_by_time_on_core_) {
+    summary += absl::StrFormat("  %s[%i] spent %.6f ms on core (%.2f%%)\n", p_stats->process_name,
```
Comment (karupayun) [inline src/OrbitGl/SchedulingStats.cpp:77]: selection time

I would like to visually differentiate a little more between Process and Threads stats.

Maybe adding a "-" in every thread stats, to make the user visualize threads as a list from process?
```
for (ThreadStats* t_stats : p_stats->thread_stats_sorted_by_time_on_core) {
      summary += absl::StrFormat("  - %s[%i] spent %.6f ms on core (%.2f%%)\n"
```

File: src/OrbitGl/SchedulingStats.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      return a->time_on_core_ns_ > b->time_on_core_ns_;
+    }
+  };
+
+  double GetTimeRangeMs() const { return time_range_ms_; }
+  uint64_t GetTimeOnCoreNs() const { return time_on_core_ns_; }
+  const std::map<int32_t, uint64_t>& GetTimeOnCoreNsByCore() const {
+    return time_on_core_ns_by_core_;
+  }
+  const std::map<int32_t, ProcessStats>& GetProcessStatsByPid() const {
```
Comment (karupayun) [inline src/OrbitGl/SchedulingStats.h:65]: I would expect this function, as well as the previous one, to be more like 
`  const ProcessStats& GetProcessStats(int32_t pid) const {
`
but feel free to disregard this comment

If you make this change, _GetTimeOnCoreNsByCore_ could be called _GetTimeOnCoreNs_

File: src/OrbitGl/SchedulingStats.h [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  };
+
+  struct ThreadStatsTimeSorter {
+    bool operator()(const ThreadStats* a, const ThreadStats* b) const {
+      return a->time_on_core_ns_ > b->time_on_core_ns_;
+    }
+  };
+
+  double GetTimeRangeMs() const { return time_range_ms_; }
+  uint64_t GetTimeOnCoreNs() const { return time_on_core_ns_; }
```
Comment (karupayun) [inline src/OrbitGl/SchedulingStats.h:61]: GetTotalTimeOnCoresNs

**Review by karupayun — APPROVED** [review]
LGTM!!! I really like these statistics by simply selecting with the right-click! :)

**Comment by pierricgimmig** [comment]
Thanks for the comments @karupayun , I will implement the suggestions.

File: src/OrbitGl/TimerTrack.cpp [hunk 6]
```diff
@@ -418,6 +418,24 @@ std::vector<std::shared_ptr<TimerChain>> TimerTrack::GetAllChains() const {
   return chains;
 }
 
+std::vector<const TextBox*> TimerTrack::GetScopesInRange(uint64_t start_ns, uint64_t end_ns) const {
+  std::vector<const TextBox*> result;
+  for (auto chain : GetAllChains()) {
+    if (chain == nullptr) continue;
+
+    for (TimerChainIterator it = chain->begin(); it != chain->end(); ++it) {
+      for (size_t k = 0; k < it->size(); ++k) {
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/OrbitGl/TimerTrack.cpp:428]: nice, thanks, done.

File: src/OrbitGl/SchedulingStats.cpp (hunk 3 above)
Comment (pierricgimmig) [reply src/OrbitGl/SchedulingStats.cpp:77]: cool, done!

File: src/OrbitGl/SchedulingStats.h (hunk 4 above)
Comment (pierricgimmig) [reply src/OrbitGl/SchedulingStats.h:65]: Oh, yeah this needs to return the map as the map contains information about the number of processes that we have stats for and this is used in the tests. I prefer leaving as is, but thanks for the comment.

File: src/OrbitGl/SchedulingStats.h (hunk 5 above)
Comment (pierricgimmig) [reply src/OrbitGl/SchedulingStats.h:61]: I prefer to leave as is for symmetry with the thread stats, I think "time on core" implies "total time on core".

File: src/OrbitGl/CaptureStats.cpp (hunk 2 above)
Comment (pierricgimmig) [reply src/OrbitGl/CaptureStats.cpp:22]: good point, thanks!

File: src/OrbitGl/CaptureStatsTest.cpp [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  scopes.push_back(CreateScope(/*pid=*/1, /*tid=*/2, /*cpu=*/7, /*start_ns=*/0, /*end_ns=*/1000));
+
+  scopes.push_back(CreateScope(/*pid=*/2, /*tid=*/3, /*cpu=*/0, /*start_ns=*/0, /*end_ns=*/100));
+  scopes.push_back(CreateScope(/*pid=*/2, /*tid=*/3, /*cpu=*/1, /*start_ns=*/0, /*end_ns=*/100));
+  scopes.push_back(CreateScope(/*pid=*/2, /*tid=*/3, /*cpu=*/2, /*start_ns=*/0, /*end_ns=*/100));
+  scopes.push_back(CreateScope(/*pid=*/2, /*tid=*/3, /*cpu=*/3, /*start_ns=*/0, /*end_ns=*/100));
+  scopes.push_back(CreateScope(/*pid=*/2, /*tid=*/3, /*cpu=*/4, /*start_ns=*/0, /*end_ns=*/100));
+  scopes.push_back(CreateScope(/*pid=*/2, /*tid=*/3, /*cpu=*/5, /*start_ns=*/0, /*end_ns=*/100));
+  scopes.push_back(CreateScope(/*pid=*/2, /*tid=*/3, /*cpu=*/6, /*start_ns=*/0, /*end_ns=*/100));
+  scopes.push_back(CreateScope(/*pid=*/2, /*tid=*/3, /*cpu=*/7, /*start_ns=*/0, /*end_ns=*/100));
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/OrbitGl/CaptureStatsTest.cpp:89]: b/186242694

File: src/OrbitGl/CaptureStatsTest.cpp (hunk 1 above)
Comment (pierricgimmig) [reply src/OrbitGl/CaptureStatsTest.cpp:97]: b/186242694

**Pull request merged**